# Download and look into the data

## Project description

The telecom operator Interconnect would like to be able to forecast their churn of clients. If it's discovered that a user is planning to leave, they will be offered promotional codes and special plan options. Interconnect's marketing team has collected some of their clientele's personal data, including information about their plans and contracts.

**Interconnect's services**

Interconnect mainly provides two types of services:

1. Landline communication. The telephone can be connected to several lines simultaneously.
2. Internet. The network can be set up via a telephone line (DSL, *digital subscriber line*) or through a fiber optic cable.

Some other services the company provides include:

- Internet security: antivirus software (*DeviceProtection*) and a malicious website blocker (*OnlineSecurity*)
- A dedicated technical support line (*TechSupport*)
- Cloud file storage and data backup (*OnlineBackup*)
- TV streaming (*StreamingTV*) and a movie directory (*StreamingMovies*)

The clients can choose either a monthly payment or sign a 1- or 2-year contract. They can use various payment methods and receive an electronic invoice after a transaction.

### Import needed libraries

In [1]:
# Data tools
import pandas as pd

from IPython.core.interactiveshell import InteractiveShell

print('Project libraries has been successfully been imported!')

Project libraries has been successfully been imported!


1.0.3  Set environment variables

In [2]:
# set to display all output not only print() or last output
InteractiveShell.ast_node_interactivity = "all"      

1.1  Open the file and look into the data.

In [3]:
try:
    contract = pd.read_csv('contract.csv')
    internet = pd.read_csv('internet.csv')
    personal = pd.read_csv('personal.csv')
    phone = pd.read_csv('phone.csv')
    
except:
    contract = pd.read_csv('/datasets/contract.csv')
    internet = pd.read_csv('/datasets/internet.csv')
    personal = pd.read_csv('/datasets/personal.csv')
    phone = pd.read_csv('/datasets/phone.csv')
    
print('Data has been read correctly!')

Data has been read correctly!


##  Data describe notes

In [4]:
# Checking 0's
def zero_check(df):
    for i in df:
        print(i,len(df[df[i]==0]))
        
# function to determine if columns in file have null values        
def get_percent_of_na(df, num):
    count = 0
    df = df.copy()
    s = (df.isna().sum() / df.shape[0])
    for column, percent in zip(s.index, s.values):
        num_of_nulls = df[column].isna().sum()
        if num_of_nulls == 0:
            continue
        else:
            count += 1
        print('{} has {} nulls, which is {:.{}%} percent of Nulls'.format(column, num_of_nulls, percent, num))
    if count != 0:
        print("\033[1m" + 'There are {} columns with NA.'.format(count) + "\033[0m")
    else:
        print()
        print("\033[1m" + 'There are no columns with NA.' + "\033[0m")       
        
# function to display general information about the dataset
def general_info(df):
    print("\033[1m" + "\033[0m")
    display(pd.concat([df.dtypes, df.count(),df.isna().sum(),df.isna().sum()*100/len(df)], keys=['type','count','na','na%'],
                      axis=1))
    print()
    print("\033[1m" + 'Head:')  
    display(df.head())
    print()
    print("\033[1m" + 'Tail:')
    display(df.tail())
    print()
    print("\033[1m" + 'Info:')
    print()
    display(df.info())
    print()
    print("\033[1m" + 'Describe:')
    print()
    display(df.describe())
    print()
    print("\033[1m" + 'Describe include: all :')
    print()
    display(df.describe(include='all'))
    print()
    print("\033[1m" + 'nulls in the columns:')
    print()
    display(get_percent_of_na(df, 4))  # check this out
    print()
    print("\033[1m" + 'Zeros in the columns:') 
    print()
    display(zero_check(df))
    print()
    print("\033[1m" + 'Shape:', df.shape)
    print()
    print()
    print('Duplicated:',"\033[1m" + 'We have {} duplicated rows\n'.format(df.duplicated().sum()) + "\033[0m")
    print()
    print("\033[1m" + 'Dtypes:')  
    display(df.dtypes)
    print()

In [5]:
print('information about the contract dataset:')
general_info(contract)

information about the contract dataset:



type  count  na  na%
customerID         object   7043   0  0.0
BeginDate          object   7043   0  0.0
EndDate            object   7043   0  0.0
Type               object   7043   0  0.0
PaperlessBilling   object   7043   0  0.0
PaymentMethod      object   7043   0  0.0
MonthlyCharges    float64   7043   0  0.0
TotalCharges       object   7043   0  0.0


Head:


customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65


Tail:


customerID   BeginDate              EndDate            Type  \
7038  6840-RESVB  2018-02-01                   No        One year   
7039  2234-XADUH  2014-02-01                   No        One year   
7040  4801-JZAZL  2019-03-01                   No  Month-to-month   
7041  8361-LTMKD  2019-07-01  2019-11-01 00:00:00  Month-to-month   
7042  3186-AJIEK  2014-08-01                   No        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
7038              Yes               Mailed check           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes           Electronic check           29.60       346.45  
7041              Yes               Mailed check           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


None


Describe:



MonthlyCharges
count     7043.000000
mean        64.761692
std         30.090047
min         18.250000
25%         35.500000
50%         70.350000
75%         89.850000
max        118.750000


Describe include: all :



customerID   BeginDate EndDate            Type PaperlessBilling  \
count         7043        7043    7043            7043             7043   
unique        7043          77       5               3                2   
top     7590-VHVEG  2014-02-01      No  Month-to-month              Yes   
freq             1         366    5174            3875             4171   
mean           NaN         NaN     NaN             NaN              NaN   
std            NaN         NaN     NaN             NaN              NaN   
min            NaN         NaN     NaN             NaN              NaN   
25%            NaN         NaN     NaN             NaN              NaN   
50%            NaN         NaN     NaN             NaN              NaN   
75%            NaN         NaN     NaN             NaN              NaN   
max            NaN         NaN     NaN             NaN              NaN   

           PaymentMethod  MonthlyCharges TotalCharges  
count               7043     7043.000000         7043  
unique                 4             NaN         6531  
top     Electronic check             NaN               
freq                2365             NaN           11  
mean                 NaN       64.761692          NaN  
std                  NaN       30.090047          NaN  
min                  NaN       18.250000          NaN  
25%                  NaN       35.500000          NaN  
50%                  NaN       70.350000          NaN  
75%                  NaN       89.850000          NaN  
max                  NaN      118.750000          NaN


nulls in the columns:


There are no columns with NA.


None


Zeros in the columns:

customerID 0
BeginDate 0
EndDate 0
Type 0
PaperlessBilling 0
PaymentMethod 0
MonthlyCharges 0
TotalCharges 0


None


Shape: (7043, 8)


Duplicated: We have 0 duplicated rows


Dtypes:


customerID           object
BeginDate            object
EndDate              object
Type                 object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
dtype: object

In [6]:
# check if there are id that return more than once
def each_id_appears_once(df):
    if len(df['customerID'].unique()) == len(df['customerID']):
        return True
    else:
        return False    

In [7]:
each_id_appears_once(contract)

True

In [8]:
contract['PaymentMethod'].unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [9]:
contract['Type'].unique()

array(['Month-to-month', 'One year', 'Two year'], dtype=object)

In [10]:
contract['PaperlessBilling'].unique()

array(['Yes', 'No'], dtype=object)

In [11]:
contract['EndDate'].unique()

array(['No', '2019-12-01 00:00:00', '2019-11-01 00:00:00',
       '2019-10-01 00:00:00', '2020-01-01 00:00:00'], dtype=object)

In [12]:
contract['BeginDate'].unique()

array(['2020-01-01', '2017-04-01', '2019-10-01', '2016-05-01',
       '2019-09-01', '2019-03-01', '2018-04-01', '2019-04-01',
       '2017-07-01', '2014-12-01', '2019-01-01', '2018-10-01',
       '2015-04-01', '2015-09-01', '2018-01-01', '2014-05-01',
       '2015-10-01', '2014-03-01', '2018-05-01', '2019-11-01',
       '2019-02-01', '2016-01-01', '2017-08-01', '2015-12-01',
       '2014-02-01', '2018-06-01', '2019-12-01', '2017-11-01',
       '2019-06-01', '2016-04-01', '2017-02-01', '2018-12-01',
       '2014-04-01', '2018-09-01', '2014-11-01', '2016-07-01',
       '2015-02-01', '2018-07-01', '2014-08-01', '2016-03-01',
       '2018-08-01', '2014-10-01', '2015-06-01', '2016-08-01',
       '2019-05-01', '2017-03-01', '2016-02-01', '2017-09-01',
       '2014-09-01', '2017-12-01', '2016-12-01', '2017-06-01',
       '2015-05-01', '2016-10-01', '2016-09-01', '2019-08-01',
       '2019-07-01', '2017-05-01', '2017-10-01', '2014-07-01',
       '2018-03-01', '2015-01-01', '2018-11-01', '2015-

In [13]:
contract['BeginDate'].value_counts()

2014-02-01    366
2019-10-01    237
2019-11-01    237
2019-09-01    237
2020-01-01    233
             ... 
2020-02-01     11
2014-01-01      7
2013-10-01      3
2013-12-01      3
2013-11-01      2
Name: BeginDate, Length: 77, dtype: int64

In [14]:
len(contract[contract['EndDate'] == 'No']) / len(contract['EndDate']) 

0.7346301292063041

### Notes about the contract data base:
In the contract we have all the customers id - 7043 total. The data is unballanced (73% not left the program)
No missing values.

Task in the preprocessing:
- Need to correct columns name to underscore lowercase.
- change BeginDate date to date type
- TotalCharges change to float64

Task in the features engineering:
- PaymentMethod, Type, PaperlessBilling with one hot encoding
- Create column based on the EndDate says if customer left or not the service.
- Check seasonality for BeginDate (month, day of the week, hour a day)
- After splitting perform class ballancing


       

In [15]:
print('information about the internet dataset:')
general_info(internet)

information about the internet dataset:



type  count  na  na%
customerID        object   5517   0  0.0
InternetService   object   5517   0  0.0
OnlineSecurity    object   5517   0  0.0
OnlineBackup      object   5517   0  0.0
DeviceProtection  object   5517   0  0.0
TechSupport       object   5517   0  0.0
StreamingTV       object   5517   0  0.0
StreamingMovies   object   5517   0  0.0


Head:


customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No


Tail:


customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
5512  6840-RESVB             DSL            Yes           No              Yes   
5513  2234-XADUH     Fiber optic             No          Yes              Yes   
5514  4801-JZAZL             DSL            Yes           No               No   
5515  8361-LTMKD     Fiber optic             No           No               No   
5516  3186-AJIEK     Fiber optic            Yes           No              Yes   

     TechSupport StreamingTV StreamingMovies  
5512         Yes         Yes             Yes  
5513          No         Yes             Yes  
5514          No          No              No  
5515          No          No              No  
5516         Yes         Yes             Yes


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB


None


Describe:



customerID InternetService OnlineSecurity OnlineBackup  \
count         5517            5517           5517         5517   
unique        5517               2              2            2   
top     7590-VHVEG     Fiber optic             No           No   
freq             1            3096           3498         3088   

       DeviceProtection TechSupport StreamingTV StreamingMovies  
count              5517        5517        5517            5517  
unique                2           2           2               2  
top                  No          No          No              No  
freq               3095        3473        2810            2785


Describe include: all :



customerID InternetService OnlineSecurity OnlineBackup  \
count         5517            5517           5517         5517   
unique        5517               2              2            2   
top     7590-VHVEG     Fiber optic             No           No   
freq             1            3096           3498         3088   

       DeviceProtection TechSupport StreamingTV StreamingMovies  
count              5517        5517        5517            5517  
unique                2           2           2               2  
top                  No          No          No              No  
freq               3095        3473        2810            2785


nulls in the columns:


There are no columns with NA.


None


Zeros in the columns:

customerID 0
InternetService 0
OnlineSecurity 0
OnlineBackup 0
DeviceProtection 0
TechSupport 0
StreamingTV 0
StreamingMovies 0


None


Shape: (5517, 8)


Duplicated: We have 0 duplicated rows


Dtypes:


customerID          object
InternetService     object
OnlineSecurity      object
OnlineBackup        object
DeviceProtection    object
TechSupport         object
StreamingTV         object
StreamingMovies     object
dtype: object

In [16]:
each_id_appears_once(internet)

True

In [17]:
# check if the internet databse id are within the contract data base

In [18]:
def is_all_id_in_sub_df_are_within_main_data_base(sub_df, main_df=contract):
    list_main = main_df['customerID'].to_list()
    list_sub = sub_df['customerID'].to_list()
    check =  all(item in list_main for item in list_sub)
    if check is True:
        print("list_main contains all elements of the list_sub")    
    else :
        print("No, List1 doesn't have all elements of the List2.")

In [19]:
is_all_id_in_sub_df_are_within_main_data_base(sub_df=internet)

list_main contains all elements of the list_sub


### Notes about Internet data base:


Here we have information about all users with Internet service. We can immediately see that 7043-5517= 1526 don't have Internet service and therefore they are not appears in this database.

Task in the preprocessing:
- Need to correct columns name to underscore lowercase.

Task in the features engineering:
- merge this data to the contract data that will become the main df
- add to InternetService column a category of - NO. This will suite the rest 1526 users that have no Internet at all. Same all the rest of the columns will receive NO for them.
- Do one hot encoding for all these features


In [20]:
print('information about the personal dataset:')
general_info(personal)

information about the personal dataset:



type  count  na  na%
customerID     object   7043   0  0.0
gender         object   7043   0  0.0
SeniorCitizen   int64   7043   0  0.0
Partner        object   7043   0  0.0
Dependents     object   7043   0  0.0


Head:


customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No


Tail:


customerID  gender  SeniorCitizen Partner Dependents
7038  6840-RESVB    Male              0     Yes        Yes
7039  2234-XADUH  Female              0     Yes        Yes
7040  4801-JZAZL  Female              0     Yes        Yes
7041  8361-LTMKD    Male              1     Yes         No
7042  3186-AJIEK    Male              0      No         No


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


None


Describe:



SeniorCitizen
count    7043.000000
mean        0.162147
std         0.368612
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000


Describe include: all :



customerID gender  SeniorCitizen Partner Dependents
count         7043   7043    7043.000000    7043       7043
unique        7043      2            NaN       2          2
top     7590-VHVEG   Male            NaN      No         No
freq             1   3555            NaN    3641       4933
mean           NaN    NaN       0.162147     NaN        NaN
std            NaN    NaN       0.368612     NaN        NaN
min            NaN    NaN       0.000000     NaN        NaN
25%            NaN    NaN       0.000000     NaN        NaN
50%            NaN    NaN       0.000000     NaN        NaN
75%            NaN    NaN       0.000000     NaN        NaN
max            NaN    NaN       1.000000     NaN        NaN


nulls in the columns:


There are no columns with NA.


None


Zeros in the columns:

customerID 0
gender 0
SeniorCitizen 5901
Partner 0
Dependents 0


None


Shape: (7043, 5)


Duplicated: We have 0 duplicated rows


Dtypes:


customerID       object
gender           object
SeniorCitizen     int64
Partner          object
Dependents       object
dtype: object

In [21]:
is_all_id_in_sub_df_are_within_main_data_base(sub_df=personal)

list_main contains all elements of the list_sub


In [22]:
each_id_appears_once(personal)

True

### Notes about personal data base:

Here we see the same users as in the contract (main). This is just addition of data so we can simply merge it to the main.


In [23]:
print('information about the phone dataset:')
general_info(phone)

information about the phone dataset:



type  count  na  na%
customerID     object   6361   0  0.0
MultipleLines  object   6361   0  0.0


Head:


customerID MultipleLines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes


Tail:


customerID MultipleLines
6356  2569-WGERO            No
6357  6840-RESVB           Yes
6358  2234-XADUH           Yes
6359  8361-LTMKD           Yes
6360  3186-AJIEK            No


Info:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


None


Describe:



customerID MultipleLines
count         6361          6361
unique        6361             2
top     5575-GNVDE            No
freq             1          3390


Describe include: all :



customerID MultipleLines
count         6361          6361
unique        6361             2
top     5575-GNVDE            No
freq             1          3390


nulls in the columns:


There are no columns with NA.


None


Zeros in the columns:

customerID 0
MultipleLines 0


None


Shape: (6361, 2)


Duplicated: We have 0 duplicated rows


Dtypes:


customerID       object
MultipleLines    object
dtype: object

In [24]:
is_all_id_in_sub_df_are_within_main_data_base(sub_df=phone)

list_main contains all elements of the list_sub


In [25]:
each_id_appears_once(phone)

True

### Notes about phone data base:
not all users have a phone line. What we can do here is to merge this database to the main data base and use OHE as so:
- Does it have only one user?
- Does it have multiple users?
If both answers be no this is someone without phone plan

# Proposed work plan

In the telecom project our goal is to develop a model that can tell is user will leave the service.
The steps to achieve that will be:

1. Download the data


2. Explore the data to determine how to treat the data in the preprocessing


3. Perform preprocessing for the data that will include
    - merge the data from all databases to one main dataframe
    - Change column name to consist format
    - convert to desired type 
    
4. Perform EDA to explore the data in depth 
 
 
5. Feature engineering 

     produce features that classification model can handle That don't create data leakage between train and test set.
    
    
6. Optional - Use unsupervised learning to see if there are clusters in the data - It can be interesting to see if the endate that are not "No" match with the clusters. If the answer to this will be yes then after split train and target to do it again for feature generation.
    
    
7. Prepare data for modeling

    - Split for tain & test in ratio of 4:1 in stratified way
    - Preform encodings (OHE and ordinal) as well as scaling since data is unbalanced
    
    
8. Model training

    - For at least 3 different models select best hyperparameters using cross-validation and validate with AUC-ROC.
    - Train models with best hyperparameters on all train set.
    - For sanity check use one dummy model


9. Model testing

    - Test each model with the test set and choose the one producing the best AUC-ROC score. 
    - If the desired story points not achieve then return to step 5 
 
 
10. Conclusions

    Draw conclusions on the model
    
Questions to the team leader
- How come there are only 4 possible contract EndDate?
       '2019-12-01', '2019-11-01', '2019-10-01', '2020-01-01'
- Please refer to my step 6. Is that something that I can do? If not how can I use unsupervised learning as a way to improve my classification model? 
    